### PR TITLE
Add CacheKeyNormalizer

### PR DESCRIPTION
**Pull request by devanych**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #60

### CODE REVIEW

**Comment by samdark** [comment]
How would you use it?

**Comment by devanych** [comment]
For example, for a composite key:

```php
$normalizer = new Yiisoft\Cache\CacheKeyNormalizer();
$key = ['part-key-1', 'part-key-2', 'part-key-3'];

/** @var Psr\SimpleCache\CacheInterface $cache  */
$cache->get($normalizer->normalize($key));
```

**Comment by samdark** [comment]
Well, I've thought about a decorator because creating a normalizer each time isn't convenient.

**Comment by samdark** [comment]
@xepozz did you mean something like in this pull requset?

**Comment by xepozz** [comment]
Yes I would like to make a separated class KeyNormalizer that will collapse a set of key into one string and replace not allowed chars with stubs.

File: tests/CacheKeyNormalizerTest.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+use function json_encode;
+use function md5;
+
+final class CacheKeyNormalizerTest extends TestCase
+{
+    private CacheKeyNormalizer $normalizer;
+
+    public function setUp(): void
+    {
+        $this->normalizer = new CacheKeyNormalizer();
```
Comment (xepozz) [inline tests/CacheKeyNormalizerTest.php:23]: Let's create it in a test case explicity

**Comment by samdark** [comment]
OK.

File: tests/CacheKeyNormalizerTest.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    public function testNormalize($key, string $excepted): void
+    {
+        $this->assertSame($excepted, $this->normalizer->normalize($key));
+    }
+
+    public function testNormalizeThrowExceptionForInvalidKey(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->normalizer->normalize($resource = fopen('php://memory', 'r'));
+        fclose($resource);
```
Comment (xepozz) [inline tests/CacheKeyNormalizerTest.php:65]:
```suggestion
        $resource = fopen('php://memory', 'r');
        $this->expectException(InvalidArgumentException::class);
        $this->normalizer->normalize($resource);
        fclose($resource);
```

**Review by xepozz — CHANGES_REQUESTED** [review]

**Review by samdark — APPROVED** [review]

File: tests/CacheKeyNormalizerTest.php (hunk 1 above)
Comment (samdark) [reply tests/CacheKeyNormalizerTest.php:23]: I think it's OK as is.

**Comment by samdark** [comment]
Thanks!

**Pull request merged**